### ISSUE TITLE
Privacy sniffs

### ISSUE THREAD

**Issue by NielsdeBlaauw** [WordPress/WordPress-Coding-Standards#1830]
## Is your feature request related to a problem?
Currently it is hard for users to find out what plugins and themes are doing with regards to data before installing. There is currently an optional ‘privacy’ section in the readme that gets put in the ‘Privacy policy postbox’. Part of this is currently described (Suggesting text for the site privacy policy)[https://developer.wordpress.org/plugins/privacy/suggesting-text-for-the-site-privacy-policy/].
## Describe the solution you’d like
We’d like to check for functions that interact with third parties or load content from external domains. Examples of this would be `wp_enqueue_script()` with an external domain, or `wp_remote_post()` which might contain user data. If we detect usage of these functions, we should check if `wp_add_privacy_policy_content()` is being called, or maybe if the readme contains a privacy policy section.

**Comment by jrfnl** [comment]
@NielsdeBlaauw Interesting idea.

I'm not sure PHPCS is the best tooling for something like this as - based on your description - this would require scanning things cross-file and looking for what's missing.

I'd be interested to hear more about how you envision this.

**Comment by NielsdeBlaauw** [comment]
Maybe we can use some of the following (peusdo code down below):

```
Sniff registers for functions that can call external resources.
External resource call found: 
    $this->phpcsFile->config['must_have_privacy_policy'] = true;
    $this->config->reporters->add( PrivacyReport )
```

```
Sniff registers for wp_add_privacy_policy_content() or any alternative
Found: 
    $this->phpcsFile->config['has_privacy_policy'] = true;
```

```
PrivacyReport:
Check config->has_privacy_policy = false  and config->must_have_privacy_policy = true.
    Output privacy policy warning
```

**Comment by jrfnl** [comment]
@NielsdeBlaauw Just checking if I understand you correctly - are you talking about using a custom PHPCS `Report` (like the existing `Full`, `Info` etc reports) ?

**Comment by NielsdeBlaauw** [comment]
Yes, but that is going to be a lot of work to be compatible with other reporting modes. If it is at all possible.

Maybe a better solution would be to add a `PrivacyFile` extending PHPCS' `DummyFile` and overwriting the `getErrors` method to check the config values during report generation.

**Comment by jrfnl** [comment]
@NielsdeBlaauw I've done some work on custom reports before and I'm not sure that's the way to go.

Reports get a notification when an error/warning is thrown and can then display those either per file and/or at the end of the run.
So they get triggered on errors/warnings thrown (but need errors/warnings to get triggered) and they can remember that information to summarize it at the end.
For an example of a custom report: https://github.com/jrfnl/QA-WP-Projects/blob/master/Reports/WPQA.php

By default, the `Full` report is shown and any other report is only shown when specifically requested, so we couldn't enforce the report to be shown to the user unless we overwrite the default `report` command-line setting which has other implications, especially when the WP standards are combined with other external standards via a custom ruleset.

> Maybe a better solution would be to add a PrivacyFile extending PHPCS' DummyFile and overwriting the getErrors method to check the config values during report generation.

That would still require a custom report to be used.

Unless there is an urgent reason you want to work on this now, let's leave this to soak for a while.
There's plenty of other open issues which could do with some attention in the mean time.
